### PR TITLE
Reduce RAM consumption

### PR DESCRIPTION
**Pull request by VeaaC**
* String table uses compact \0 terminated string storage. Sadly this requires unsafe code
* Tag table uses i40 instead of i64 to save 9 bytes per entry

In total this reduces the maximum amount of memory allocated for the compilation of Europe from 14.9GB down to 9.5 GB

### CODE REVIEW

File: osmflatc/src/strings.rs [hunk 1]
```diff
@@ -1,9 +1,63 @@
 use ahash::AHashMap;
-use inlinable_string::{InlinableString, StringExt};
+
+#[derive(Debug, Clone, Copy)]
+struct TerminatedStringPtr {
```
Comment (boxdot) [inline osmflatc/src/strings.rs:4]: Sad, that `&CStr` is 16 bytes

File: osmflatc/src/strings.rs [hunk 2]
```diff
@@ -1,9 +1,63 @@
...
+    fn borrow(&self) -> &[u8] {
+        self.as_bytes()
+    }
+}
 
 #[derive(Debug, Default)]
 pub struct StringTable {
-    indexed_data: AHashMap<InlinableString, u64>,
+    // Append only, we will never reallocate any data inside
+    data: Vec<Vec<u8>>,
```
Comment (boxdot) [inline osmflatc/src/strings.rs:54]: You could store Box<[u8]> in the outer `Vec`, which tells you that you can't change it anymore. It is also smaller than Vec by one word (no capacity).

See https://doc.rust-lang.org/std/vec/struct.Vec.html#method.into_boxed_slice

**Review by boxdot — APPROVED** [review]

File: osmflatc/src/strings.rs (hunk 2 above)
Comment (VeaaC) [reply osmflatc/src/strings.rs:54]: Yes, but that saves only 8 bytes per 4 MB.

File: osmflatc/src/strings.rs (hunk 1 above)
Comment (VeaaC) [reply osmflatc/src/strings.rs:4]: Yes, I was looking to CStr as an abstraction.

Best case I really like to use less than 8 bytes (5 are enough): if I could just store an index into `data`, and the Hashmap would provide `data` as a `context` to Hash/PartialEq...

**Pull request merged**